name: Acceptance Tests
on:
  push:
    branches: main
  pull_request: {}

jobs:
  give-fixtures-a-chance-to-run:
    name: Give Fixtures A Chance To Run
    runs-on: ubuntu-latest
    steps:
      - run: sleep 20

  handles-waiting-for-pull-request-target-from-pull-request:
    name: Can handle a pull_request_target event from a pull request event
    runs-on: ubuntu-latest
    if: ${{ github.event_name == 'pull_request' }}
    needs: give-fixtures-a-chance-to-run
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Wait for pull_request_target fixture
        uses: ./
        with:
          workflows: pull-request-target-success-fixture.yml
          wait-interval-seconds: 5
          wait-max-seconds: 60
          repo-token: ${{ github.token }}
          debug: "on"

  with-succeeding-workflow:
    name: Should Succeed (waiting for succeeding workflow)
    runs-on: ubuntu-latest
    needs: give-fixtures-a-chance-to-run
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Wait for success-fixture
        uses: ./
        with:
          workflows: success-fixture.yml
          wait-interval-seconds: 5
          wait-max-seconds: 60
          repo-token: ${{ github.token }}

  with-two-succeeding-workflows:
    name: Should Succeed (waiting for two succeeding workflows)
    runs-on: ubuntu-latest
    needs: give-fixtures-a-chance-to-run
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Wait for success-fixture
        uses: ./
        with:
          workflows: |
            success-fixture.yml
            fast-success-fixture.yml
          wait-interval-seconds: 5
          wait-max-seconds: 60
          repo-token: ${{ github.token }}

  workflow-doesnt-exist:
    name: Should Fail (waiting for non existent workflow)
    runs-on: ubuntu-latest
    needs: give-fixtures-a-chance-to-run
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Wait for non-existent workflow
        continue-on-error: true
        id: test
        uses: ./
        with:
          workflows: non-existent.yml
          wait-interval-seconds: 5
          wait-max-seconds: 60
          repo-token: ${{ github.token }}

      - name: Assert error occurred
        if: ${{ steps.test.outcome != 'failure' }}
        run: echo "This should have failed" && exit 1

  workflow-run-failed:
    name: Should Fail (waiting for failing workflow)
    runs-on: ubuntu-latest
    needs: give-fixtures-a-chance-to-run
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Wait for failing workflow
        continue-on-error: true
        id: test
        uses: ./
        with:
          workflows: failure-fixture.yml
          wait-interval-seconds: 5
          wait-max-seconds: 60
          repo-token: ${{ github.token }}

      - name: Assert error occurred
        if: ${{ steps.test.outcome != 'failure' }}
        run: echo "This should have failed" && exit 1

  workflow-succeeded-in-the-past:
    name: Should Fail (waiting for workflow that isn't running now but did in the past)
    runs-on: ubuntu-latest
    needs: give-fixtures-a-chance-to-run
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Wait for failing workflow
        continue-on-error: true
        id: test
        uses: ./
        with:
          # This was an old name for a workflow, as can be
          # seen by it not running anymore we name-changed it
          # but it can be used for this purpose
          workflows: fixture.yml
          wait-interval-seconds: 5
          wait-max-seconds: 60
          repo-token: ${{ github.token }}

      - name: Assert error occurred
        if: ${{ steps.test.outcome != 'failure' }}
        run: echo "This should have failed" && exit 1

  one-failed-one-succeeded-success-first:
    name: Should Fail (one failed and one succeeded with success first)
    runs-on: ubuntu-latest
    needs: give-fixtures-a-chance-to-run
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Wait for one failing and one succeeding workflow with success first
        continue-on-error: true
        id: test
        uses: ./
        with:
          workflows: |
            fast-success-fixture.yml
            failure-fixture.yml
          wait-interval-seconds: 5
          wait-max-seconds: 60
          repo-token: ${{ github.token }}

      - name: Assert error occurred
        if: ${{ steps.test.outcome != 'failure' }}
        run: echo "This should have failed" && exit 1

  one-failed-one-succeeded-failure-first:
    name: Should Fail (one failed and one succeeded with failure first)
    runs-on: ubuntu-latest
    needs: give-fixtures-a-chance-to-run
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Wait for one failing and one succeeding workflow with failure first
        continue-on-error: true
        id: test
        uses: ./
        with:
          workflows: |
            success-fixture.yml
            fast-failure-fixture.yml
          wait-interval-seconds: 5
          wait-max-seconds: 60
          repo-token: ${{ github.token }}

      - name: Assert error occurred
        if: ${{ steps.test.outcome != 'failure' }}
        run: echo "This should have failed" && exit 1

  two-failed:
    name: Should Fail (waiting on two failed)
    runs-on: ubuntu-latest
    needs: give-fixtures-a-chance-to-run
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Wait for two failing
        continue-on-error: true
        id: test
        uses: ./
        with:
          workflows: |
            fast-failure-fixture.yml
            failure-fixture.yml
          wait-interval-seconds: 5
          wait-max-seconds: 60
          repo-token: ${{ github.token }}

      - name: Assert error occurred
        if: ${{ steps.test.outcome != 'failure' }}
        run: echo "This should have failed" && exit 1

  logging-on:
    name: With logging on
    runs-on: ubuntu-latest
    needs: give-fixtures-a-chance-to-run
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Wait with logging on
        uses: ./
        with:
          workflows: fast-success-fixture.yml
          wait-interval-seconds: 5
          wait-max-seconds: 60
          repo-token: ${{ github.token }}
          debug: "on"

  logging-verbose:
    name: With logging verbose
    runs-on: ubuntu-latest
    needs: give-fixtures-a-chance-to-run
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Wait with logging on
        uses: ./
        with:
          workflows: fast-success-fixture.yml
          wait-interval-seconds: 5
          wait-max-seconds: 60
          repo-token: ${{ github.token }}
          debug: verbose
